sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/PeriodicMetricReader.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.export;
+
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.internal.DaemonThreadFactory;
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Wraps a {@link MetricExporter} and automatically reads and exports the metrics every export
+ * interval. Metrics may also be dropped when it becomes time to export again, and there is an
+ * export in progress.
+ */
+public class PeriodicMetricReader implements MetricReader {
+  private static final Logger logger = Logger.getLogger(PeriodicMetricReader.class.getName());
+
+  /**
+   * Builds a factory that will register and start a PeriodicMetricReader.
+   *
+   * <p>This will export once every 5 minutes.
+   *
+   * <p>This will spin up a new daemon thread to schedule the export on.
+   *
+   * @param exporter The exporter receiving metrics.
+   */
+  public static MetricReaderFactory create(MetricExporter exporter) {
+    return create(exporter, Duration.ofMinutes(5));
+  }
+
+  /**
+   * Builds a factory that will register and start a PeriodicMetricReader.
+   *
+   * <p>This will spin up a new daemon thread to schedule the export on.
+   *
+   * @param exporter The exporter receiving metrics.
+   * @param duration The duration (interval) between metric export calls.
+   */
+  public static MetricReaderFactory create(MetricExporter exporter, Duration duration) {
+    return create(
+        exporter,
+        duration,
+        Executors.newScheduledThreadPool(1, new DaemonThreadFactory("PeriodicMetricReader")));
+  }
+
+  /**
+   * Builds a factory that will register and start a PeriodicMetricReader.
+   *
+   * @param exporter The exporter receiving metrics.
+   * @param duration The duration (interval) between metric export calls.
+   * @param scheduler The service to schedule export work.
+   */
+  public static MetricReaderFactory create(
+      MetricExporter exporter, Duration duration, ScheduledExecutorService scheduler) {
+    return new PeriodicMetricReaderFactory(exporter, duration, scheduler);
+  }
+
+  private final MetricProducer producer;
+  private final MetricExporter exporter;
+  private final ScheduledExecutorService scheduler;
+  private final Scheduled scheduled;
+  private final Object lock = new Object();
```

Comment:
field declarations should go above static methods. You've been poisoned by @anuraaga !

Reply:
Ha! I miss Scala where field declarations are the constructor.